adding boilerplate code

cmbrennan002 opened the pull request:
---

# Description

This takes code to search twitter from the repo asf_online_data_exploration

Fixes # (issue) 01

# Instructions for Reviewer

You will need to add the access token to your .env file - will send to you separately

Then, you can run:

`cd pipeline
python recent_search_twitter.py`

Based on current search parameters, that should take 10 tweets (the min number we can search for) that have come from betway and contain the word 'promotion' in the last seven days, and add it to the json file stored in the s3 bucket.

cmbrennan002 committed:
adding boilerplate code

cmbrennan002 commented on ds_digital_ads/config/data_collection_parameters.py:
@india-kerle this is the file we can update with our search queries. For now I've just done a toy one

india-kerle replied:
okay great - very helpful. I would put this in `utils` instead of config personally

india-kerle replied:
and perhaps rename to data_collection_utils to include anything else that would be helpful for this?

cmbrennan002 replied:
Great! There's already a data_collection_utils.py, so have put it in there and might move some of what's there to the getters

cmbrennan002 commented on ds_digital_ads/pipeline/recent_search_twitter.py:
```diff
@@ -0,0 +1,293 @@
+                json_response, max_ids_json, query_tag, date_time_collection_start
+            )
+
+        # collect tweets from next pages
+        """
+        while "next_token" in json_response["meta"]:
+            query_params["next_token"] = json_response["meta"]["next_token"]
+
+            json_response = connect_to_endpoint(headers, query_params)
+            data = process_twitter_data(json_response, data)
+        """
```
I have commented this out to prevent us collecting too many tweets while we're testing

india-kerle replied:
smart !! I usually use a # to comment out parts of code - there's a fancy short cut to hash tons of lines of code in your IDE (command then /)

india-kerle commented on ds_digital_ads/config/base.yaml:
```diff
@@ -0,0 +1,6 @@
+# bucket info
+S3_BUCKET: "ds-digital-ads"
```
oh nice i usually add this to __init__ and import from there - i think its 6 and 1 half dozen of the other really

india-kerle replied:
i.e. `from ds_digital_ads import BUCKET_NAME`

cmbrennan002 replied:
I like it this way too! Have edited

india-kerle commented on ds_digital_ads/config/data_collection_parameters.py:
```diff
@@ -0,0 +1,20 @@
+"""
+Script with parameters for data collection.
```
```suggestion
Utils for twitter data collection.
```

india-kerle commented on ds_digital_ads/config/data_collection_parameters.py:
```diff
@@ -0,0 +1,20 @@
+    "place.fields": "id,country,country_code,name,full_name,geo,place_type,contained_within",
+    "media.fields": "media_key,type,url,duration_ms,preview_image_url,public_metrics,alt_text",
+    "expansions": "author_id,geo.place_id,attachments.media_keys",
+    "max_results": 10,  # number of results per page when making a call to the API
```
woof so many options! I need to go through the documentation !

india-kerle commented on ds_digital_ads/pipeline/recent_search_twitter.py:
```diff
@@ -0,0 +1,293 @@
+"""
+Script to collect Twitter data from the past 7 days using the recent search endpoint, given a set of rules and query parameters.
+If data on a specific ruleset has been collected sometime in the past 7 days, only new data gets collected.
```
I always find it helpful to add a line about how to run the script at the top if i ever need to rerun

```suggestion
Script to collect Twitter data from the past 7 days using the recent search endpoint, given a set of rules and query parameters.
If data on a specific ruleset has been collected sometime in the past 7 days, only new data gets collected.

python ds_digital_ads/pipeline/recent_search_twitter.py
```

cmbrennan002 replied:
Love this, I will do it in other places too! Thanks for the suggestion :)

india-kerle commented on ds_digital_ads/pipeline/recent_search_twitter.py:
```diff
@@ -0,0 +1,293 @@
+
+ENDPOINT_URL = "https://api.twitter.com/2/tweets/search/recent"
+S3_BUCKET = base_config["S3_BUCKET"]
+DATA_COLLECTION_FOLDER = base_config["RECENT_SEARCH_TWITTER_S3_DATA_COLLECTION_FOLDER"]
```
maybe we could add this to the data_collection_utils.py?

india-kerle commented on ds_digital_ads/pipeline/recent_search_twitter.py:
```diff
@@ -0,0 +1,293 @@
+    return data
+
+
+def get_max_ids_json(s3_bucket, folder) -> dict:
```
nice to define the types of these inputs

india-kerle commented on ds_digital_ads/pipeline/recent_search_twitter.py:
```diff
@@ -0,0 +1,293 @@
+
+def update_max_ids_json(
+    json_response, max_ids_json, query_tag, date_time_collection_start
+):
```
ditto define types here

india-kerle commented on ds_digital_ads/pipeline/recent_search_twitter.py:
```diff
@@ -0,0 +1,293 @@
+if __name__ == "__main__":
+    load_dotenv()
+    bearer_token = os.environ.get("BEARER_TOKEN")
+
```
lets add some kind of error handling if bearer_token is None

india-kerle commented on ds_digital_ads/pipeline/recent_search_twitter.py:
```diff
@@ -0,0 +1,293 @@
+    max_ids_json[query_tag]["collection_datetime"] = date_time_collection_start
+
+
+def collect_and_process_twitter_data(
+    bearer_token: str,
+    rules: list,
+    query_params: dict,
+    folder: str,
+    s3_bucket: str = None,
+):
```
interesting - i would usually just have this all after if __name__ == "__main__": and not as a function in and of itself

india-kerle replied:
it could be nice to add some logging as things happen but that's largely because i love to log

cmbrennan002 replied:
Both of these are very fair! Since we're just re-using the code for EDA, for the sake of time, I've left as-is - but let me know if you have a strong preference!

india-kerle commented on ds_digital_ads/utils/data_collection_utils.py:
```diff
@@ -0,0 +1,76 @@
+"""
+Utility functions for data collection.
+"""
```
ooooh interesting - in previous repos, i usually put these in a `data_getters.py` in the `getters` directory but to be fair i guess they're not reeeeally getters

india-kerle replied:
https://github.com/nestauk/dap_prinz_green_jobs/blob/dev/dap_prinz_green_jobs/getters/data_getters.py - the load_to_s3 and save_to_s3 functions are nice because they handle a ton of different format types - could be good to use these to enable flexibility if we end up wanting to save other formats when we analyse the data downstream.

cmbrennan002 replied:
Great call! Have moved these to getters (otherwise the utils files is a bit of a jumble, since we moved what was in config into there). I don't have a strong preference but I like the delineation. Have also copied across the generic functions

india-kerle commented on requirements.txt:
```diff
@@ -0,0 +1,2 @@
+boto3
+pandas
```
need to also install:

```
python-dotenv
```

cmbrennan002 replied:
Thanks! Have added now

india-kerle reviewed (commented):
Looks good! I needed to pip install `python-dotenv` in order to run the script but had no worries after that.

cmbrennan002 committed:
addressing PR comments

cmbrennan002 commented:
@india-kerle thanks heaps for the thoughtful PR! I've addressed most comments (agreed with all your structural suggestions) and flagged a few functions to leave for now. I've tested it runs on my machine, but given we've shifted a few things around, would really appreciate it if you could re-test on yours. Then let me know and will merge :)

india-kerle commented:
cool! Will take a look now :) @cmbrennan002

india-kerle commented:
ran like a dream! feel free to merge @cmbrennan002
